ovn-tester/cms/openstack/tests/base_openstack.py

```diff
@@ -0,0 +1,28 @@
+import logging
+
+from ovn_ext_cmd import ExtCmd
+from ovn_context import Context
+from ovn_workload import ChassisNode
+
+from cms.openstack import OpenStackCloud
+
+log = logging.getLogger(__name__)
+
+
+class BaseOpenstack(ExtCmd):
+    def __init__(self, config, central_node, worker_nodes, global_cfg):
+        super().__init__(config, central_node, worker_nodes)
+
+    def run(self, ovn: OpenStackCloud, global_cfg):
+        # create ovn topology
+        worker_count = len(ovn.worker_nodes)
+        with Context(ovn, "base_cluster_bringup", worker_count) as ctx:
+            for i in ctx:
+                worker_node: ChassisNode = ovn.worker_nodes[i]
+                log.info(
+                    f"Provisioning {worker_node.__class__.__name__} "
+                    f"({i+1}/{worker_count})"
+                )
+                worker_node.provision(ovn)
+
+            _ = ovn.new_project(gw_nodes=ovn.worker_nodes)
```

Comment:
I like this, it uses the same approach as the `base_clustrer_bringup` test that every ovn-kubernetes test case runs to initialize things before running other tests.

The `new_project` method should probably be called N times per test scenario where N is input from the test-scenario configuration without direct relation to the number of workers configured. Perhaps a `n_projects` key under `base_openstack`?

About the number of `gw_nodes`. Neutron currently has a hard coded maximum of 5 GW nodes per router, so we should probably mimic that behavior. The nodes are either selected by chance or by load (load as in number of routers hosted at any given priority)

Reply:
That's a good point about the GW nodes, thank you. I didn't know about the gw limit per router. It should be fairly easy to distribute chassis in balanced way

Reply:
I had to eat my words about the easy redistribution mechanism :D balance distribution of individual GW priorities between bunch of routers turned out rather complex so I defaulted to using `random.sample()` :D